service/history/hsm/tree.go

```diff
@@ -601,3 +752,138 @@ func GenerateEventLoadToken(event *historypb.HistoryEvent) ([]byte, error) {
 	}
 	return proto.Marshal(ref)
 }
+
+// markDeleted marks a single node as deleted
+func (n *Node) markDeleted(vt *persistencespb.VersionedTransition) error {
+	if n.persistence.Deleted {
+		return fmt.Errorf("node already deleted: %v", n.Key)
+	}
+
+	update := stateUpdate{
+		persistence: &persistencespb.StateMachineNode{
+			Deleted:                       true,
+			LastUpdateVersionedTransition: vt,
+			TransitionCount:               n.persistence.TransitionCount + 1,
+			Children:                      n.persistence.Children,
+			Data:                          n.persistence.Data,
+			InitialVersionedTransition:    n.persistence.InitialVersionedTransition,
+		},
+		cache: &cachedMachine{
+			dataLoaded: n.cache.dataLoaded,
+			data:       n.cache.data,
+			children:   n.cache.children,
+			dirty:      true,
+		},
+		deleted: true,
+	}
+
+	return n.applyStateUpdate(update)
+}
+
+func (n *Node) applyStateUpdate(update stateUpdate) error {
+	n.Lock()
+	defer n.Unlock()
+
+	if update.persistence != nil {
+		n.persistence = update.persistence
+	}
+	if update.cache != nil {
+		n.cache = update.cache
+	}
+	if update.dirty {
+		n.cache.dirty = true
+	}
+	if update.deleted {
+		n.persistence.Deleted = true
+	}
+
+	return nil
+}
+
+// applyOperation applies a single operation during sync
+func (n *Node) applyOperation(ctx context.Context, op *persistencespb.StateMachineOperation) error {
+	switch op.OperationType {
+	case persistencespb.StateMachineOperation_OPERATION_TYPE_DELETE:
+		n.persistence.Deleted = true
+		n.persistence.LastUpdateVersionedTransition = op.VersionedTransition
+		n.persistence.TransitionCount++
+		n.cache.dirty = true
+
+		root := n.GetRoot()
+		return root.opLog.LogOperation(ctx, op)
+
+	case persistencespb.StateMachineOperation_OPERATION_TYPE_TRANSITION:
```

Comment:
You'll want to log the transition operation too when a machine transitions.
Essentially you can get rid of the "outputs" concept and replace it with the op log.